gh-103740: Improve support of __slots__ for variable-length classes (#141636)
python/cpython: Doc/reference/datamodel.rst

  descriptor directly from the base class). This renders the meaning of the
  program undefined.  In the future, a check may be added to prevent this.

* :exc:`TypeError` will be raised if *__slots__* other than *__dict__* and
  *__weakref__* are defined for a class derived from a
  :c:member:`"variable-length" built-in type <PyTypeObject.tp_itemsize>`,
  unless the C type is defined with :c:macro:`Py_TPFLAGS_ITEMS_AT_END`.

>   For example, *__slots__* other than *__dict__* and *__weakref__* cannot
>   be defined on subclasses of :class:`int`, :class:`bytes`, or :class:`tuple`.

Author: Maybe use "they"?
```suggestion
  For example, they cannot be defined on subclasses of
  :class:`int`, :class:`bytes`, or :class:`tuple`.
```

  descriptor directly from the base class). This renders the meaning of the
  program undefined.  In the future, a check may be added to prevent this.

* :exc:`TypeError` will be raised if *__slots__* other than *__dict__* and
  *__weakref__* are defined for a class derived from a
  :c:member:`"variable-length" built-in type <PyTypeObject.tp_itemsize>`
  without the :c:macro:`Py_TPFLAGS_ITEMS_AT_END` flag such as

>   :class:`int`, :class:`bytes`, and :class:`tuple`.

Reviewer: Make it clearer that these are examples of types that cause failure, not examples of classes that use `Py_TPFLAGS_ITEMS_AT_END`.

I find it hard to say this clearly *and* succinctly; let's favour clarity:

```suggestion
* :exc:`TypeError` will be raised if *__slots__* other than *__dict__* and
  *__weakref__* are defined for a class derived from a
  :c:member:`"variable-length" built-in type <PyTypeObject.tp_itemsize>`,
  unless the C type is defined with :c:macro:`Py_TPFLAGS_ITEMS_AT_END`.
  For example, *__slots__* other than *__dict__* and *__weakref__* cannot
  be defined on subclasses of :class:`int`, :class:`bytes`, or :class:`tuple`.
```

Reviewer: Or maybe something like:
```suggestion
* :exc:`TypeError` will be raised if *__slots__* are defined for a class
  derived from a
  :c:member:`"variable-length" built-in type <PyTypeObject.tp_itemsize>`
  (such as :class:`int`, :class:`bytes`, and :class:`tuple`), except if:
  
  * *__slots__* only contains *__dict__* and/or *__weakref__*,
    or is empty; or
  * the class uses the C flag :c:macro:`Py_TPFLAGS_ITEMS_AT_END`.
```

  of the iterator's values. However, the *__slots__* attribute will be an empty
  iterator.

.. versionchanged:: next
   Allowed defining the *__dict__* and *__weakref__* *__slots__* for any class.

> 

Reviewer: I think this would work better as a single paragraph:
```suggestion
```
